core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -983,8 +979,14 @@ protected void onException(Channel channel, Exception e) throws IOException {
      */
     protected abstract void closeChannels(List<Channel> channel) throws IOException;
 
-
-    protected abstract void sendMessage(Channel channel, BytesReference reference, Runnable sendListener) throws IOException;
+    /**
+     * Sends message to channel. The listener will be called with an exception if the send operation throws an exception. If the operation
+     * is successful, the listener will be called with null.
```

Comment:
I created a version of this PR with an ActionListener and then backed it out and just used a consumer.

The prior version of this code only used a Runnable and that Runnable was always called (success or failure). It seems like the value of that callback was primarily to release resources (bytes).

But there was some tension with that and the fact that ping success/failure used this callback as its execution mechanism. And that the TransportServiceAdapter uses the callback to ensure that things like onResponseSent are called (we actually still call that success or fail with this PR. I'm not sure if we want to change that and only call success?).

Anyway - I moved to a Consumer for the following reasons:
1. An ActionListener would prevent lambdas from being used and require implementing two methods and releasing resources in both methods
2. I felt that the main reason for this callback was to release resources. Only the ping required different behavior for success/failure.
3. I also kind of felt like the ping success/failure logic and TransportServiceAdapter logic should be called from the netty pipeline and the callback should exclusively exist for releasing resources.

But then I guess if number 3 is the correct approach maybe this PR should just do that and leave the resource releasing callback as a Runnable?